Upgrade react-day-picker to v7

Skovy opened the pull request:
#### Fixes #2054

#### Changes proposed in this pull request:

Upgrade the `react-day-picker` dependency from version `^5.3.0` to `^7.0.7` and update to accomadate breaking changes in the `react-day-picker` package highlighted in their [CHANGELOG](https://github.com/gpbl/react-day-picker/blob/master/CHANGELOG.md).

#### Reviewers should focus on:

Type accuracy and breaking changes highlighted in the `react-day-picker` [CHANGELOG](https://github.com/gpbl/react-day-picker/blob/master/CHANGELOG.md).

palantirtech commented:
Thanks for your interest in [palantir/blueprint](https://github.com/palantir/blueprint), @Skovy! Before we can accept your pull request, you need to sign our contributor license agreement - just visit https://cla.palantir.com/ and follow the instructions. Once you sign, I'll automatically update this pull request.

Skovy commented:
Getting issues using submodule exports. Not sure the best steps to resolve, because it looks [`react-day-picker` does not export `DayPickerProps`](https://github.com/gpbl/react-day-picker/blob/master/types/index.d.ts)

```
@blueprintjs/datetime: Submodule import paths from this package are disallowed; import from the root instead (no-submodule-imports)
@blueprintjs/datetime:    8 | import * as moment from "moment";
@blueprintjs/datetime:    9 | import * as React from "react";
@blueprintjs/datetime: > 10 | import { DayPickerProps } from "react-day-picker/types/props";
```

1. Can I disable the rule for these lines using a disable comment?
2. Open a PR for [`react-day-picker`](https://github.com/gpbl/react-day-picker) to add these to the index export?

adidahiya commented:
@Skovy you can add an exception to this list in the tslint config: https://github.com/palantir/blueprint/blob/8f05618ddcdc04f243c198f276d18873d3cfef4e/packages/tslint-config/index.js#L54

add `"react-day-picker/types"` to the array.

giladgray commented on packages/datetime/src/dateInput.tsx:
```diff
@@ -7,7 +7,7 @@
 import * as moment from "moment";
 import * as React from "react";
-import * as ReactDayPicker from "react-day-picker";
+import { DayPickerProps } from "react-day-picker/types/props";
```
is it possible to import this without a submodule? like from `react-day-picker` itself like it used to be? (i haven't poked at the actual package yet)

giladgray replied:
on further examination, it seems not. 👍

giladgray replied:
in this case, @Skovy please add the tslint-config exception as suggested by @adidahiya above.

Skovy replied:
👍 yea, I was looking into that as well and was disappointed only `DayPicker` was exported.

cmslewis replied:
Filed: https://github.com/gpbl/react-day-picker/issues/617

giladgray commented:
@Skovy nice! tests failures seem to be relevant -- DayPicker-related. check 'em out 🔨

Skovy commented:
@giladgray yea noticed that. I can dig into that, definitely seems related to the breaking changes in the CHANGELOG, just trying to understand the specs, abstractions and how to debug 😋 

Any useful docs in the repo for good debugging techniques?

giladgray commented:
`yarn test:karma:debug` in the datetime package will run the tests in watch mode.

And `yarn dev:datetime` in the Root directory will launch the dev server of the docs site which reloads on changes.

Happy hunting!

Skovy commented on packages/datetime/src/dateRangePicker.tsx:
```diff
@@ -225,11 +227,11 @@ export class DateRangePicker extends AbstractPureComponent<IDateRangePickerProps
             locale,
             localeUtils,
             modifiers,
+            showOutsideDays: true,
```
Curious if this prop should be configurable given it's always hidden since this PR https://github.com/palantir/blueprint/pull/586/files#r98813760?

cmslewis replied:
We can leave this as is for now. Seems like changing it here would be out of scope for this PR.

Skovy commented on packages/datetime/src/_daterangepicker.scss:
```diff
@@ -14,6 +14,10 @@
 
+  .DayPicker-NavButton--interactionDisabled {
+    display: none;
+  }
```
Breaking change in v6.0.0. The element is now always rendered, but with this CSS class conditionally: https://github.com/gpbl/react-day-picker/blob/master/CHANGELOG.md#v600-2017-06-16

Skovy commented on packages/datetime/test/dateRangePickerTests.tsx:
```diff
@@ -406,8 +406,10 @@ describe("<DateRangePicker>", () => {
-            assert.lengthOf(document.getElementsByClassName("DayPicker-NavButton--next"), 0);
+            // react-day-picker still renders the navigation but with a interaction disabled class
+            assert.lengthOf(document.getElementsByClassName("DayPicker-NavButton--prev"), 1);
+            assert.lengthOf(document.getElementsByClassName("DayPicker-NavButton--next"), 1);
```
Not sure if there is a better way to actually assert visibility of these elements, given they are now always rendered.

cmslewis replied:
Do we need to reach into the `style` of each element to check the `visibility`?

giladgray replied:
what about looking for that conditional class?

giladgray replied:
or, if they're always there, just remove these assertions. no need to assert on external library behavior. case in point: when that behavior changes and breaks your tests (but our code is still fine).

cmslewis commented on packages/datetime/src/datePicker.tsx:
```diff
@@ -7,7 +7,9 @@
 import * as classNames from "classnames";
 import * as React from "react";
-import * as ReactDayPicker from "react-day-picker";
+import { default as ReactDayPicker } from "react-day-picker";
```
Whoa, funky syntax. Haven't seen this before, but it's interesting. If it's equivalent, should we just do:
```tsx
import ReactDayPicker from "react-day-picker";
```
?

_Source: https://stackoverflow.com/a/31092287/5199574_

adidahiya replied:
yes

Skovy replied:
It would have to be `import DayPicker from "react-day-picker";`, since the default export is `DayPicker` (not `ReactDayPicker`). However, that conflicts with the actual `DayPicker` component name. So used this approach to rename `DayPicker` to `ReactDayPicker`. Open to other alternatives, but as far as I know, the above recommendation won't work?

adidahiya replied:
> It would have to be import DayPicker from "react-day-picker";, since the default export is DayPicker (not ReactDayPicker)

Are you sure about this? `default` exports are (by design) anonymous, so you can import them with whatever name you want.

Skovy replied:
@adidahiya will that work with the defined types though? 

https://github.com/gpbl/react-day-picker/blob/3156e3de1664caa15cf12cfd0fdd96350874e8c5/types/index.d.ts#L4

giladgray replied:
it will. default exports don't have a name of their own. you can name them whatever you want with the `import DefaultName from "package"` syntax.

Skovy replied:
Wow. I swear I tried that 🤔 TIL. Appreciate the feedback! 😊

giladgray commented:
@Skovy a few small comments still need to be addressed (esp. the default import thing), then this will be good to go!

Skovy commented:
@giladgray appreciate all the feedback. Updated the default imports and removed some of the redundant assertions in the specs. Please take another look 👍

Skovy commented:
Not sure if this is related or a flake?

```
FAILED TESTS:
    ✖ "after each" hook
      PhantomJS 2.1.1 (Linux 0.0.0)
    TypeError: undefined is not an object (evaluating 'wrapper.assertIsOpen') (test/index.ts:102819)
```

giladgray commented:
@Skovy when was the last time you merged `develop` into this branch? (our latest `develop`, not your fork's.) done a lot of work this week to address failures like these.

Skovy commented:
Hey all, curious when the next beta will be released containing this PR?

giladgray commented:
@Skovy tomorrow!
